lib/utils/prompt/context_reader_test.go

```diff
@@ -186,6 +186,70 @@ func TestContextReader_ReadPassword(t *testing.T) {
 	})
 }
 
+func TestNotifyExit_restoresTerminal(t *testing.T) {
+	oldStdin := Stdin()
+	t.Cleanup(func() { SetStdin(oldStdin) })
+
+	pr, _ := io.Pipe()
+
+	devNull, err := os.OpenFile(os.DevNull, os.O_RDWR, 0666)
+	require.NoError(t, err, "Failed to open %v", os.DevNull)
+	defer devNull.Close()
+
+	term := &fakeTerm{reader: pr}
+	ctx := context.Background()
+
+	tests := []struct {
+		name        string
+		doRead      func(ctx context.Context, cr *ContextReader) error
+		wantRestore bool
+	}{
+		{
+			name: "no pending read",
+			doRead: func(ctx context.Context, cr *ContextReader) error {
+				<-ctx.Done()
+				return ctx.Err()
+			},
+		},
+		{
+			name: "pending clean read",
+			doRead: func(ctx context.Context, cr *ContextReader) error {
+				_, err := cr.ReadContext(ctx)
+				return err
+			},
+		},
+		{
+			name: "pending password read",
+			doRead: func(ctx context.Context, cr *ContextReader) error {
+				_, err := cr.ReadPassword(ctx)
+				return err
+			},
+			wantRestore: true,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			term.restoreCalled = false // reset state between tests
+
+			cr := NewContextReader(pr)
+			cr.term = term
+			cr.fd = int(devNull.Fd()) // arbitrary
+			SetStdin(cr)
+
+			ctx, cancel := context.WithCancel(ctx)
+			go func() {
+				time.Sleep(1 * time.Millisecond) // give the read time to block
```

Comment:
Why do we need this wait? One millisecond feels like it could be flaky, if we absolutely must hit the block before canceling the context.

Reply:
Also, isn't this just a convoluted way to do `context.WithTimeout`?

Reply:
It shouldn't flake, but I'd like to give it time to block without having to assume much about how the reads are implemented.

Good point about context.WithTimeout, changed.